From e1bcceef54d73f50e9177ea1effa7a9b1cb616a6 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Fri, 20 Apr 2018 14:12:42 -0700
Subject: [PATCH] joyent/node-sshpk-agent#16 Ed25519 cert logic should handle
 mixed sshpk versions in the same process Reviewed by: Robert Mustacchi
 <rm@joyent.com>

---
 lib/client.js | 26 +++++++++++++++++++++-----
 package.json  |  2 +-
 2 files changed, 22 insertions(+), 6 deletions(-)

diff --git a/lib/client.js b/lib/client.js
index 91d8d57..f3f0e84 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -376,6 +376,8 @@ Client.prototype.createCertificate =
  * it's not even documented, let alone used anywhere else).
  */
 function certToBuffer(cert, k) {
+	assert.ok(sshpk.Key.isKey(k, [1, 2]));
+
 	var buf = sshpk.Certificate.formats.openssh.toBuffer(cert);
 	var sshbuf = new SSHBuffer({ buffer: buf });
 	var type = sshbuf.readString();
@@ -398,12 +400,26 @@ function certToBuffer(cert, k) {
 		break;
 	case 'ed25519':
 		/*
-		 * For some reason the public key gets encoded again for
-		 * ed25519 certs. The mysteries will never cease.
+		 * Our source key can be from a different sshpk version to the
+		 * one we actually are using. ed25519 support was added in the
+		 * [1,4] Key version, and the format for storage changed in
+		 * [1,6] (sshpk 1.14.0).
+		 *
+		 * OpenSSH expects us to prepend the public key to the private
+		 * key (transmitting it twice), as this was the original format
+		 * that the ed25519 reference impl used. sshpk used it as well
+		 * until 1.14.0 when it was changed to make PEM/x509 support for
+		 * ed25519 simpler.
 		 */
-		sshbuf.writeBuffer(k.part.A.data);
-		var edk = Buffer.concat([k.part.A.data, k.part.k.data]);
-		sshbuf.writeBuffer(edk);
+		if (sshpk.Key.isKey(k, [1, 6])) {
+			sshbuf.writeBuffer(k.part.A.data);
+			var edk = Buffer.concat([k.part.A.data, k.part.k.data]);
+			sshbuf.writeBuffer(edk);
+		} else {
+			assert.ok(sshpk.Key.isKey(k, [1, 4]));
+			sshbuf.writeBuffer(k.part.R.data);
+			sshbuf.writeBuffer(k.part.r.data);
+		}
 		break;
 	default:
 		throw (new Error('Key type ' + k.type + ' not supported'));
diff --git a/package.json b/package.json
index 622d4df..824ea35 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "sshpk-agent",
-  "version": "1.7.0",
+  "version": "1.7.1",
   "description": "ssh-agent client for use with sshpk",
   "main": "lib/index.js",
   "scripts": {
-- 
2.21.0

